### DISCUSSION TITLE
Executing code in memory

### DISCUSSION THREAD

**Discussion by kevinboone** [rsta2/circle#349] [Q&A]
Hi folks

In some circumstances it would be useful if Circle applications could load code from SD card into RAM and execute it. Then large, rarely-used pieces of the application need not necessarily be loaded with the kernel. 

At present, though, I believe that execution of code from RAM is prevented. At least, I see a comment in one of the test samples to that effect. Can this restriction (if I have understood correctly) be turned off in some way?

An additional complication, if I understand correctly, is that the memory maps seem to be different for the different Pi models. I don't know if this is something that is under the control of Circle, or whether it is imposed by the processor or hardware. 

I'm also aware that loading code and running it is treading on the toes of what a full operating system would do, and may be completely outside the design scope of Circle.

Comments welcome.
Thanks, Kevin

**Comment by rsta2** [comment]
Yes, execution of code above `_etext` is normally prohibited. This makes debugging more easy, because one gets an exception, when the CPU is jumping into the wild. [Here](https://github.com/rsta2/circle/issues/282#issuecomment-1072769330) you can find info, how to disable this.

The memory map of Circle is defined in the files *include/circle/memorymap[64].h* and described [here](https://github.com/rsta2/circle/blob/master/doc/memorymap.txt). If you `#define MEM_HEAP_START 0x1000000` for example, you should have an unused memory space between 0x900000 and 0xFFFFFF on all Raspberry Pi models.

Rene

**Reply by kevinboone** [reply]
Thanks. I wonder -- could the code running in that area above 0x900000 call into the 'kernel' so ask it (the 'kernel') to call malloc() to allocate more memory? That would be read/write memory, not executable memory. I'm trying to think of a way that the loaded code could get working memory, given that the different Pi boards have different memory maps. The code loaded into memory won't know where the memory allocation function is in the kernel, but I guess the kernel could store an address some place above 0x9000000, that the loaded code could call into. 

That sounds like it should work, but comments are welcome. Do the CPUs in the Pi boards require that branch instructions use the least-significant bit of the address to specify whether the jump is to Arm or Thumb code? I remember struggling with this on other ARM-based boards.

**Reply by rsta2** [reply]
I did read [here](https://forums.raspberrypi.com/viewtopic.php?p=1531456#p1531456), that it is possible to link your own code, which is placed at 0x900000, against the kernel*.elf file, which is produced by the normal Circle build process. This way you should be able to call `malloc()` (or anything, which is in the kernel image) as it is normally done.

All Circle code is ARM (not Thumb) code. If your own code is ARM code too, you do not have to care about switching between the modes.

**Reply by stephaneweg** [reply]
Thank, it works for me too
